src/models/session.rs

```diff
@@ -0,0 +1,290 @@
+use crate::schema::sessions;
+use chrono::NaiveDateTime;
+use diesel::prelude::*;
+use ipnetwork::IpNetwork;
+use sha2::digest::consts::U32;
+use sha2::digest::generic_array::GenericArray;
+use sha2::{Digest, Sha256};
+use std::borrow::Cow;
+use std::net::IpAddr;
+
+const TOKEN_LENGTH: usize = 32;
+
+#[derive(Clone, Debug, PartialEq, Eq, Identifiable, Queryable)]
+#[table_name = "sessions"]
+pub struct Session {
+    pub id: i32,
+    pub user_id: i32,
+    hashed_token: Vec<u8>,
+    pub created_at: NaiveDateTime,
+    pub last_used_at: NaiveDateTime,
+    pub revoked: bool,
+    last_ip_address: IpNetwork,
+    pub last_user_agent: String,
+}
+
+impl Session {
+    /// Creates a new session builder that can be used to insert new session
+    /// into the database.
+    ///
+    /// ```
+    /// let session = Session::new()
+    ///    .user_id(user.id)
+    ///    .token(&token)
+    ///    .last_ip_address(ip_addr)
+    ///    .last_user_agent(user_agent)
+    ///    .build()?
+    ///    .insert(&conn)?;
+    /// ```
+    ///
+    /// New tokens can be generated by using `Session::generate_token()`.
+    pub fn new() -> NewSessionBuilder<'static> {
+        NewSessionBuilder::default()
+    }
+
+    /// Looks for an unrevoked session with the given `token` in the database
+    /// and returns `Some(Session)` if successful.
+    ///
+    /// If the session exists then the `last_used_at`, `last_ip_address` and
+    /// `last_user_agent` fields will be updated in the same step.
+    pub fn find_by_token_and_update(
+        conn: &PgConnection,
+        token: &str,
+        ip_address: IpAddr,
+        user_agent: &str,
+    ) -> Result<Option<Self>, diesel::result::Error> {
+        let hashed_token = Self::hash_token(token);
+
+        let sessions = sessions::table
+            .filter(sessions::revoked.eq(false))
+            .filter(sessions::hashed_token.eq(hashed_token.as_slice()));
+
+        // If the database is in read only mode, we can't update these fields.
+        // Try updating in a new transaction, if that fails, fall back to reading
+        conn.transaction(|| {
+            diesel::update(sessions)
+                .set((
+                    sessions::last_used_at.eq(diesel::dsl::now),
+                    sessions::last_ip_address.eq(IpNetwork::from(ip_address)),
+                    sessions::last_user_agent.eq(user_agent),
+                ))
+                .get_result(conn)
+                .optional()
+        })
+        .or_else(|_| sessions.first(conn).optional())
```

Comment:
It's probably a good idea to do the update only if:

 - The IP address has changed, or
 - The User-Agent has changed, or
 - The last_used_at has increased by more than 15 minutes.
 
 This will save you from having to do a DB write on every cookie-authenticated request.